aptos-move/aptos-intent/src/codegen.rs

```diff
@@ -0,0 +1,489 @@
+};
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{PartialVMError, PartialVMResult},
+    file_format::*,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, TypeTag},
+    metadata::Metadata,
+    transaction_argument::TransactionArgument,
+    vm_status::StatusCode,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Default)]
+struct Context {
+    address_pool: BTreeMap<AccountAddress, AddressIdentifierIndex>,
+    identifier_pool: BTreeMap<Identifier, IdentifierIndex>,
+    module_pool: BTreeMap<(AddressIdentifierIndex, IdentifierIndex), ModuleHandleIndex>,
+    function_pool: BTreeMap<(ModuleHandleIndex, IdentifierIndex), FunctionHandleIndex>,
+    struct_pool: BTreeMap<(ModuleHandleIndex, IdentifierIndex), StructHandleIndex>,
+    signature_pool: BTreeMap<Signature, SignatureIndex>,
+    script: CompiledScript,
+    args: Vec<Vec<u8>>,
+    ty_args: Vec<TypeTag>,
+    returned_val_to_local: Vec<u16>,
+    args_to_local: Vec<u16>,
+
+    return_counts: Vec<u16>,
+    signer_counts: u16,
+    locals: Vec<SignatureToken>,
+    parameters: Vec<SignatureToken>,
+    ty_args_to_idx: BTreeMap<TypeTag, u16>,
+}
+
+fn instantiate(token: &SignatureToken, subst_mapping: &BTreeMap<u16, u16>) -> SignatureToken {
+    use SignatureToken::*;
+
+    match token {
+        Bool => Bool,
+        U8 => U8,
+        U16 => U16,
+        U32 => U32,
+        U64 => U64,
+        U128 => U128,
+        U256 => U256,
+        Address => Address,
+        Signer => Signer,
+        Vector(ty) => Vector(Box::new(instantiate(ty, subst_mapping))),
+        Struct(idx) => Struct(*idx),
+        StructInstantiation(idx, struct_type_args) => StructInstantiation(
+            *idx,
+            struct_type_args
+                .iter()
+                .map(|ty| instantiate(ty, subst_mapping))
+                .collect(),
+        ),
+        Reference(ty) => Reference(Box::new(instantiate(ty, subst_mapping))),
+        MutableReference(ty) => MutableReference(Box::new(instantiate(ty, subst_mapping))),
+        TypeParameter(idx) => TypeParameter(*subst_mapping.get(idx).unwrap()),
+    }
+}
+
+impl Context {
+    fn import_address(
+        &mut self,
+        module: &CompiledModule,
+        idx: AddressIdentifierIndex,
+    ) -> PartialVMResult<AddressIdentifierIndex> {
+        self.import_address_by_name(module.address_identifier_at(idx))
+    }
+
+    fn import_address_by_name(
+        &mut self,
+        address: &AccountAddress,
+    ) -> PartialVMResult<AddressIdentifierIndex> {
+        if let Some(result) = self.address_pool.get(address) {
+            return Ok(*result);
+        }
+        if self.script.address_identifiers.len() >= TableIndex::MAX as usize {
+            return Err(PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS));
+        }
```

Comment:
Yeah, the logic here assumes that if something is not in the hashmap then it also does not exist in the script. Fine if we're starting from an empty script, but personally I'd like to see this becoming a general library we can use to modify `CompiledModule` and `CompiledScript`